docs/2022-04-07-Full-Corectness-All-Path-Reachability-Proofs.md

```diff
@@ -0,0 +1,412 @@
+Proving Full Correctness All-Path Reachability Claims
+=====================================================
+
+This document details Full Correctness All-Path Reachability without solving the
+most-general-unifier (MGU) problem.
+MGU will be detailed in a separate document.
+
+_Prepared by Traian Șerbănuță, based on [proving All-Path Reachability
+claims](2019-03-28-All-Path-Reachability-Proofs.md)_
+
+Definitions
+-----------
+
+### All-path finally
+
+Given a formula `φ`, let `♢φ` denote the formula “all-path finally” `φ`,
+defined by:
+
+```
+♢φ := μX.φ ∨ (○X ∧ •⊤)
+```
+
+one consequence of the above is that `♢φ ≡ φ ∨ (○♢φ ∧ •⊤)`.
```

Comment:
In our papers, `≡` means the same thing as `:=`, meaning that the LHS is defined as an abbreviation of the RHS. But here I guess you mean `|- ♢φ = φ ∨ (○♢φ ∧ •⊤)`, which is saying that the equation is provable. 

How do you prove it? It's not obvious to me. By unfolding the fixpoint, I can prove
```
|- ♢φ = φ \/ •♢φ
```
But I don't see how that RHS is equal to `φ ∨ (○♢φ ∧ •⊤)`.

Reply:
OK, maybe I'm missing the point here, but if I consider the "functor" defined on sets by
`F(X) := φ ∨ (○X ∧ •⊤)`, which would read as 
`F(X)` is the union between the set of configuration for which `φ` holds and those which are not stuck and whose all possible transitions lead into `X`.

Then wouldn't `LFP(F)` satisfy that `LFP(F) = F(LPF(F))`? If so, this should be enough to establish the equality you are wondering about.
Moreover, isn't is right that `LFP(F)` can be set-theoretically defined as `⋁ₙFⁿ(⊥)`
Then noting that `F(⊥) = φ ∨ (○⊥ ∧ •⊤) = φ ∨ (¬•¬⊥ ∧ •⊤) = φ ∨ (¬•⊤ ∧ •⊤) = φ ∨ ⊥ = φ` is the set of configurations for which `φ` holds, we would have that
- `F(F(⊥))` is the union between the set of configuration for which `φ` holds and those which are not stuck and whose all possible transitions lead into configurations for which `φ` holds.
- `F(F(F(⊥)))` is the union between the set of configuration for which `φ` holds and those which are not stuck and whose all possible transitions lead into the union between the set of configuration for which `φ` holds and those which are not stuck and whose all possible transitions lead into configurations for which `φ` holds.
- ...

Hence `Fⁿ(⊥)` seems to define the set of configurations for which  `φ` holds on all paths in at most `n-1` steps.

Since by the above expression for `LPF(F)`, a configuration is in `LFP(F)` if there exists a natural number such that the configuration belongs to `Fⁿ(⊥)`, this says that `LFP(F)` defines the set of configuration for which  `φ` holds on all paths after a finite number of steps, which is in my opinion total correctness.

If we agree on this, I'll try to rewrite the existing text to incorporate this argument.

Reply:
about `≡`: I was using it to denote semantic equivocalence; I've now replaced it with plain equality in the document.

Reply:
>Then wouldn't LFP(F) satisfy that LFP(F) = F(LPF(F))? If so, this should be enough to establish the equality you are wondering about.

Yes, I misunderstood it with the usual "eventually". It is correct that we can prove `|- ♢φ = φ ∨ (○♢φ ∧ •⊤)`

Btw, I highly suggest we use a different symbol than ♢. ♢ has already been used to represent (one-path) eventually, which is defined as `♢phi = mu X . phi \/ •X`. 

>  Isn't is right that LFP(F) can be set-theoretically defined as ⋁ₙFⁿ(⊥)

No, it is incorrect. See transfinite induction (https://en.wikipedia.org/wiki/Transfinite_induction). 

Basically, ⋁ₙFⁿ(⊥) (which is a limit) may not converge. We need to compute F(⋁ₙFⁿ(⊥)) (which is F applied to the limit) and keep applying F again, until we consume all the natural numbers (again) and reach to another limit, and then apply F to the second limit. And keep repeating until finally, it converges. 

Therefore, the only rigorous way (in my view) to reason about the semantics of a fixpoint is to following the semantics of matching logic and the Knaster-Tarski theorem. Unfolding can give us some intuition, but it won't give us a rigorous proof (unless we use transfinite unfolding). 

However, in this case, I think if we can prove the following using the matching logic proof system, then we establish the correctness of the encoding thoroughly:

```
(Goal) |- LHS = RHS
where LHS = mu X.φ ∨ (○X ∧ •⊤)
and RHS = (nu X.φ ∨ (○X ∧ •⊤)) /\ (mu X . ○X)
```

Reply:
Thanks for pointing that out. It skipped my mind that F would need to be continuous.

I think the proposal would still work as-is (and as further detailed in the new commits) if `F` can be shown to be Omega-continuous, i.e, that it commutes with limits of increasing chains.  As far as I know, most of the commonly used `F`s have this property; I'll try to think whether it's something we can generally show for the kind of patterns we use here.